pkg/services/ngalert/api/api_ruler.go

```diff
@@ -636,5 +637,50 @@ func calculateChanges(ctx context.Context, ruleStore store.RuleStore, groupKey n
 	}, nil
 }
 
+// calculateAutomaticChanges scans all affected groups and creates either a noop update that will increment the version of each rule as well as re-index other groups.
+// this is needed to make sure that there are no any concurrent changes made to all affected groups.
+// Returns a copy of changes enriched with either noop or group index changes for all rules in
+func calculateAutomaticChanges(ch *changes) *changes {
+	updatingRules := make(map[ngmodels.AlertRuleKey]struct{}, len(ch.Delete)+len(ch.Update))
+	for _, update := range ch.Update {
+		updatingRules[update.Existing.GetKey()] = struct{}{}
+	}
+	for _, del := range ch.Delete {
+		updatingRules[del.GetKey()] = struct{}{}
+	}
+	var toUpdate []ruleUpdate
+	for groupKey, rules := range ch.AffectedGroups {
+		if groupKey != ch.GroupKey {
+			ngmodels.RulesGroup(rules).SortByGroupIndex()
+		}
+		idx := 1
+		for _, rule := range rules {
+			if _, ok := updatingRules[rule.GetKey()]; ok { // exclude rules that are going to be either updated or deleted
+				continue
+			}
+			upd := ruleUpdate{
+				Existing: rule,
+				New:      rule,
+			}
+			if groupKey != ch.GroupKey {
+				if rule.RuleGroupIndex != idx {
+					upd.New = ngmodels.CopyRule(rule)
+					upd.New.RuleGroupIndex = idx
```

Comment:
Do we need to do this if the group is already sorted by index? What is this solving:
1. Deduplicating any duplicated indexes. Is that something that can actually happen?
2. Making sure there are no "holes" in the rule group index sequence. Is that important? Is it ok to have a sparse index sequence?

Is there another case I'm missing?

Reply:
For example, 

| Group Index |1 |2 | 3 | 4 |
| -- | - | - | - | - |
|Group1| A |B | C |D |
|Group2| E | F | G | H |

User moves rule `G` to Group1, i.e. submits a POST request `Group1: A, B, C, G, D`. 

This code re-indexes Group2 to eliminate gaps